daemon/src/request/waiting_room.rs

```diff
@@ -128,6 +104,48 @@ where
     }
 }
 
+/// A transition in the waiting room and any result the waiting room produced
+#[derive(Debug, PartialEq)]
+pub struct TransitionWithResult<T, R> {
+    /// The event that caused the state change
+    event: Event,
+    /// The time  the change occurred
+    timestamp: T,
+    /// The state before the change
+    state_before: HashMap<Revision, SomeRequest<T>>,
+    /// The state after the change
+    state_after: HashMap<Revision, SomeRequest<T>>,
```

Comment:
My reasoning here is that the purpose of this feature is to help with debugging so the less logic involved in producing transitions the better. Just copying the raw before and after state means that we can be relatively sure that it's a faithful representation of the data the program is working against.

I don't _think_ this should have serious performance implications. The waiting room is generally a reasonably small hashmap and production of these events is relatively infrequent - although as I say this it occurs to me that this may not be true for seeds.

Reply:
> My reasoning here is that the purpose of this feature is to help with debugging so the less logic involved in producing transitions the better.

Ya, I can see why that would be useful. My worry would be that it's conflating some debugging and the waiting room's logic. I'll always have the before and after states for debugging, because I have to make the call on it and I get the result.

You're right that the waiting room is generally small and it wouldn't impact it greatly. I'm not sure if a seed would need this code, but you never know :)

So maybe we should look for a way to separate out what the waiting room is doing and ways of tracing the state of affairs. I'm thinking something like:
```rust
pub struct Traced<T, U> {
  before: T,
  after: U,
}

impl<T, U> Traced {
  /// Convert a `T` into a `U` with the supplied function,
  /// but keep track of the before and after states.
  pub trace<T: Clone, U>(t: T, f: impl FnOnce(T) -> U) -> Traced<T, U> {
    let before = t.clone()
    Traced { before, after: f(t) }
  }
}
```

The only thing I'm not sure about here is the higher-order function. It would need to be polymorphic in owned, `&`, and `&mut` but maybe we just write a load of boiler-plate `trace`, `trace_ref`, and `trace_mut` :upside_down_face: 

What do you think of this? Do you see another way of separating the concerns?

Reply:
I see what you're getting at. My main concern when implementing the `state_before` and `state_after` assignments in the waiting room was to avoid exposing the waiting room internals to the rest of the application. However, the fact that the `HashMap<PeerId, SomeRequest<T>>` type is currently exposed in the `WaitingRoomTransition` event emitted by the subroutines somewhat defeats that objective. I think there are two approaches:

1. Expose the `state_before` and `state_after` in the event as `serde_json::Value`s and leave the actual content of the up to the waiting room. This achieves the original implementation-hiding intent and would justify leaving the actual production of those fields within the waiting room. It does make consuming events nicely more complicated because you don't really know what's in them, but as this is primarily for debugging that may not be a concern.
2. Perform the event generation in the subroutines as using your `Traced` type in combination with a helper method to remove some of the repetitiveness. This would probably also mean moving the `Event` enum out of `waiting_room` and into `subroutines`.

I think I'm inclined to the latter approach.

Reply:
Ya, to be honest, I think 2. sounds more reasonable to me. It just seems like there's a layering approach where the waiting room is the pure logic, then there's an execution layer in the `subroutines`, and we fit a piece in-between that helps with debugging and tracing. 

What might help the debugging component is having the `WaitingRoom` give enough information about the transitions so that these can be easily convert to `Event`s, if that makes sense?

Reply:
> What might help the debugging component is having the WaitingRoom give enough information about the transitions so that these can be easily convert to Events, if that makes sense?

Are you referring here to being able to produce whatever the `before` and `after` fields of `Traced` should be for some given type?

Reply:
No sorry, I was referring to providing the caller with checking the `state` of the `Request`. I'm not sure how "inspectable" it currently is so that you can generate the `Event` :)